front: add, get and delete paced trains

SharglutDev opened the pull request:
See commits (will be merged all together)

part of #10615

> [!NOTE]
> PR is now testable !

As some logic from train schedule has been reused for the pace train mode, be sure to check that all previous behavior when paced train mode is off are still working (add/edit/get/delete, duplicate, import/export, nge, drag, selecting/projecting a train, stdcm)

E2E Tests : 
- part of add : https://github.com/osrd-project/osrd-tests/issues/89 (7-8-10-11 parts)
- duplicate/delete : https://github.com/osrd-project/osrd-tests/issues/92

SharglutDev committed:
front: handle add, get, delete and duplicate paced trains

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: convert paced train time range duration and cadence to Duration

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -0,0 +1,52 @@
+
+import type { ValidConfig } from '../types';
+
+export default function formatTimetableItemPayload<T extends boolean>(
+  validConfig: ValidConfig,
+  { isPacedTrainMode }: { isPacedTrainMode: T }
+): T extends true ? PacedTrainBase : TrainScheduleBase {
```
Hm, we do really need generics here? Could we instead have a function to format the common fields, and then separate functions for each type which call the common function?

SharglutDev replied:
Simplified the function a bit. Let me know if it's ok for you :)

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -0,0 +1,52 @@
+  { isPacedTrainMode }: { isPacedTrainMode: T }
+): T extends true ? PacedTrainBase : TrainScheduleBase {
+  const {
+    baseTrainName,
```
It seems to me like destructuring isn't buying us much here… The function would be half shorter without it, WDYT?

SharglutDev replied:
No strong opinion between destructuring all or repeating valid config each time :) I changed it.

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Why do we need to disable this one? Which variables are unused?

SharglutDev replied:
Remnant of the original function I copied. Removed it for both.

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/AddTrainScheduleButton.tsx:
```diff
@@ -51,21 +52,81 @@ const AddTrainScheduleButton = ({
+          upsertTrainSchedules([formattedNewTrainSchedule]);
+        }
+      } catch (e) {
+        setIsWorking(false);
```
To avoid repetition, this could be moved to a `finally` block.

SharglutDev replied:
Done

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/AddTrainScheduleButton.tsx:
```diff
@@ -51,21 +52,81 @@ const AddTrainScheduleButton = ({
+            })
+          );
+          setIsWorking(false);
+          dtoImport();
```
To avoid repetition, `dtoImport()` could be moved outside of the `if`/`else`.

SharglutDev replied:
Done

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/checkCurrentConfig.ts:
```diff
@@ -134,27 +137,34 @@ const checkCurrentConfig = (
+  if (showPacedTrains) {
+    // Prevent to block the train creation if a paced train field is invalid but we want to add a train schedule
+    if (isPacedTrainMode) {
+      if (cadence.total('minute') < 1) {
```
Nit: adding `.total('minute')` belongs to the previous commit, where the type of `cadence` is changed. (TypeScript missed this because `Duration` has a `valueOf()` method, which makes comparison operators work between `Duration` values… maybe it wasn't such a bright idea to implement this method…)

SharglutDev replied:
Should be fixed :)

emersion reviewed (commented):
I've looked at the first two commits! Will continue the review later.

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -49,7 +58,15 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
       { timetableId: scenario?.timetable_id },
       {
-        skip: !scenario,
+        skip: !scenario || showPacedTrains,
```
Hum I think you should fetch the paced trains even if we need to show the paced trains no ?
```suggestion
        skip: !scenario,
```

SharglutDev replied:
What was I thinking 🙃

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -58,6 +75,9 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
     [fetchedTrainSchedulesResults]
   );
 
+  const pacedTrainsResults = useMemo(() => fetchedPacedTrains || [], [fetchedPacedTrains]);
```
Hm you can replace pacedTrainResults by these formatted trains, and this way reuse it below, for instance in useEffects (same for trainScheduleResults)
```suggestion
  const pacedTrains: PacedTrainResultWithPacedTrainId[] = useMemo(
    () =>
      showPacedTrains
        ? (fetchedPacedTrains || []).map((trainSchedule) => ({
            ...trainSchedule,
            id: formatEditoastTrainIdToPacedTrainId(trainSchedule.id),
          }))
        : [],
    [showPacedTrains, fetchedPacedTrains]
  );
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -122,17 +169,42 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
+
+  useEffect(() => {
+    // TODO Paced trains : remove the if and extra depth in https://github.com/OpenRailAssociation/osrd/issues/10791
+    if (showPacedTrains) {
+      const formattedRawTrainSchedules: TrainScheduleResultWithTrainId[] =
+        trainSchedulesResults.map((trainSchedule) => ({
+          ...trainSchedule,
+          id: formatEditoastTrainIdToTrainScheduleId(trainSchedule.id),
+        }));
+      const formattedRawPacedTrains: PacedTrainResultWithPacedTrainId[] = pacedTrainsResults.map(
+        (trainSchedule) => ({
+          ...trainSchedule,
+          id: formatEditoastTrainIdToPacedTrainId(trainSchedule.id),
+        })
+      );
+      const sortedTimatableItems = sortBy(
+        [...formattedRawTrainSchedules, ...formattedRawPacedTrains],
+        'start_time'
+      );
+      setTimetableItems(sortedTimatableItems);
+    }
```
If you have this above:
```
  const trainSchedules: TrainScheduleResultWithTrainId[] = useMemo(
    () =>
      (fetchedTrainSchedulesResults || []).map((trainSchedule) => ({
        ...trainSchedule,
        id: formatEditoastTrainIdToTrainScheduleId(trainSchedule.id),
      })),
    [fetchedTrainSchedulesResults]
  );
```
then, you can simply do (pacedTrains will be empty if !showPacedTrains): 
```suggestion
      const sortedTimetableItems = sortBy([...trainSchedules, ...pacedTrains], 'start_time');
      setTimetableItems(sortedTimetableItems);
    }
```

Then, you can remove the previous useEffect

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -12,30 +12,39 @@ import {
+  // TODO Paced trains : remove these 3 states in https://github.com/OpenRailAssociation/osrd/issues/10791
   const [trainSchedules, setTrainSchedules] = useState<TrainScheduleResultWithTrainId[]>();
-  const [trainIdsToFetch, setTrainIdsToFetch] = useState<TrainId[]>();
-  const [trainIdsToProject, setTrainIdsToProject] = useState<Set<TrainId>>(new Set());
+  const [trainIdsToFetch, setTrainIdsToFetch] = useState<TimetableItemId[]>();
+  const [trainIdsToProject, setTrainIdsToProject] = useState<Set<TimetableItemId>>(new Set());
```
You can remove trainSchedules and trainIdsToFetch and directly use timetableItems
```suggestion
  const [trainIdsToProject, setTrainIdsToProject] = useState<Set<TimetableItemId>>(new Set());
```

SharglutDev replied:
Done. Made me change some parts of "not showPacedTrains" part but seems to be still working :)

Uriel-Sautron commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +59,179 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
   );
 
-  const trainsIds = useMemo(() => trainSchedulesResults.map((t) => t.id), [trainSchedulesResults]);
+  const trainScheduleIds = useMemo(
+    () => (fetchedTrainSchedulesResults || []).map((t) => t.id),
+    [fetchedTrainSchedulesResults]
+  );
```
```suggestion
  const trainScheduleIds = useMemo(
    () => (fetchedTrainSchedulesResults || []).map((trainSchedule) => trainSchedule.id),
    [fetchedTrainSchedulesResults]
  );
```

emersion commented on front/src/utils/trainId.ts:
```diff
@@ -11,6 +17,10 @@ export const isOccurrence = (id: string): id is OccurrenceId => {
   id.startsWith('trainschedule-');
 
+export const isPacedTrainResultWithPacedTrainId = (
+  timetableITem: TimetableItemWithTimetableId
```
Typo: "timetableItem"

SharglutDev replied:
:ninja:

emersion commented on front/src/utils/trainId.ts:
```diff
@@ -11,6 +17,10 @@ export const isOccurrence = (id: string): id is OccurrenceId => {
 
+export const isPacedTrainResultWithPacedTrainId = (
+  timetableITem: TimetableItemWithTimetableId
+): timetableITem is PacedTrainResultWithPacedTrainId => 'paced' in timetableITem;
```
I'd slightly prefer to check `isPacedTrain(timetableItem.id)` here

SharglutDev replied:
Done

emersion commented on front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -0,0 +1,125 @@
+            duration: Duration.parse(formattedItem.paced.duration),
+            step: Duration.parse(formattedItem.paced.step),
+          },
+        } as PacedTrainWithDetails;
```
Instead of casting, could we set the type of the variable?

```ts
const formattedPacedTrains: PacedTrainWithDetails = …
```

SharglutDev replied:
You're right, done.

emersion commented on front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -0,0 +1,125 @@
+        } as PacedTrainWithDetails;
+        acc[1].push(formattedPacedTrains);
+      } else {
+        acc[0].push(formattedItem as TrainScheduleWithDetails);
```
Is this cast really necessary?

SharglutDev replied:
It's not, you're right.

emersion commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -44,6 +45,34 @@ const osrdEditoastApi = generatedEditoastApi
+          const result: PacedTrainResult[] = [];
+          while (!reachEnd) {
+            const promise = dispatch(
+              // TODO : update the endpoint with the proper one when back is ready
```
Maybe this TODO can be dropped now?

SharglutDev replied:
Good catch

emersion commented on front/tests/utils/paced-train.ts:
```diff
@@ -0,0 +1,45 @@
+  return responseData;
+}
+
+export const getOccurrenceName = (pacedTrainName: string, index: number): string => {
```
Hm, doesn't this duplicate `useOccurrences()`? If we're duplicating code, we're not testing anything. I'd prefer to hardcode a list of expected names in an array instead.

SharglutDev replied:
Asked to @Maymanaf before doing that but he preferred not to depend on an external util (but I get what you would say that). I can make your change, but it will be a lot since I'll have to hard code for all paced trains, all the expected occurrence names.
Can't we consider that the current state of this function is the one we want and if somehow the one in osrd changes, it will break the tests here ?

SharglutDev replied:
After some discussions, I'll reduce the amount of occurrences and add the hardcoded names, start time and arrival time

SharglutDev replied:
Done

emersion commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
```diff
@@ -371,14 +402,101 @@ class ScenarioTimetableSection extends CommonPage {
+      this.occurrenceItem.nth(occurrenceIndex).locator('.occurrence-item-name')
+    ).toHaveText(computedName);
+
+    const occurrenceStartTime = dayjs(start_time)
+      .add(occurrenceIndex * formattedStep, 'minutes')
+      .format('HH:mm');
```
Same here, we're not testing anything if we are copying over the logic to compute the start time of each occurrence.

SharglutDev replied:
Ok so you'ld prefer I hardcode all expected start time here as well ?

SharglutDev replied:
Done

emersion commented on front/tests/008-timetable-item.spec.ts:
```diff
@@ -242,4 +242,19 @@ test.describe('Verify train schedule elements and filters', () => {
+    );
+
+    // Verify paced train item
+    for (let i = 0; i < pacedTrainsData.length; i += 1) {
```
This can be simplified with `for (const pacedTrain of pacedTrainsData)`

SharglutDev replied:
Wanted to do that first but I also need the index I'm looping on.

emersion replied:
Indeed!

emersion commented on front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -0,0 +1,125 @@
+        ...otherProps,
+      };
+
+      if (isPacedTrainResultWithPacedTrainId(formattedItem)) {
```
This passes a partially constructed train result to the function. In particular, `formattedItem` will not yet have a `paced` field, so the function will always return false here?

Seems safer to check the `timetableItem.id` instead of operating on an incomplete value.

SharglutDev replied:
With the change in this [comment](https://github.com/OpenRailAssociation/osrd/pull/10966#discussion_r1989447261), it should be fine I think ?

emersion replied:
Yeah, it should work, but is still passing a partially constructed object to the function. What I mean is that it's invalid to have an object with a paced train ID but without a `paced` field. Using the object before it's fully constructed can lead to surprises, especially for functions accepting both train schedules and paced trains.

Not a big deal though, I'm fine with the status quo.

emersion replied:
Unfortunately this ended up being a pretty deep rabbit hole, and fixing all of the type issues with my suggestion would result in a lot less readable code. Let's keep it that way for now, even if the types aren't perfectly correct.

emersion commented on front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -0,0 +1,125 @@
+  );
+
+  const [trainScheduleWithDetails, pacedTrainWithDetails] = relevantTimetableItems.reduce<
+    [TrainScheduleWithDetails[], PacedTrainWithDetails[]]
```
This `reduce()` is tricky to read for me. I would prefer a `for` loop.

SharglutDev replied:
:smiling_face_with_tear:

clarani replied:
🥲

emersion replied:
The `reduce()` has been dropped now \o/

emersion commented on front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -0,0 +1,125 @@
+import { isScheduledPointsNotHonored, isTooFast } from '../utils';
+
+/**
+ * This function formats the timetable items with their simulation summaries.
+ * Mapping the simulation results to the timetable items needs to be done in a second part
+ * because otherwise typescript doesn't understand that a PacedTrainWithDetails will necessarily
+ * be mapped with a PacedTrainId and not a TrainScheduleId (same for TrainScheduleWithDetails)
```
Hm, curious. The following snippet works for me:

<details>

```ts
const formatTimetableItemSummaries = (
  timetableItemIds: TimetableItemId[],
  rawSummaries: Map<TimetableItemId, SimulationSummaryResult>,
  rawTimetableItems: Map<TimetableItemId, TimetableItemWithTimetableId>,
  rollingStocks: LightRollingStockWithLiveries[]
): Map<TimetableItemId, TimetableItemWithDetails> => {
  const relevantTimetableItems = compact(
    timetableItemIds.map((timetableItemId) => rawTimetableItems.get(timetableItemId))
  );

  const items = relevantTimetableItems.map((timetableItem): TimetableItemWithDetails | null => {
    const rollingStock = rollingStocks.find((rs) => rs.name === timetableItem.rolling_stock_name);

    const timetableItemSummary = rawSummaries.get(timetableItem.id);

    if (!timetableItemSummary) return null;

    let notHonoredReason: TimetableItemWithDetails['notHonoredReason'];
    if (timetableItemSummary.status === 'success') {
      if (isTooFast(timetableItem, timetableItemSummary)) notHonoredReason = 'trainTooFast';
      if (isScheduledPointsNotHonored(timetableItem, timetableItemSummary))
        notHonoredReason = 'scheduleNotHonored';
    }

    const startTime = new Date(timetableItem.start_time);

    const otherProps =
      timetableItemSummary.status === 'success'
        ? {
            isValid: true,
            arrivalTime: new Date(startTime.getTime() + timetableItemSummary.time),
            duration: new Duration({ milliseconds: timetableItemSummary.time }),
            pathLength: formatKmValue(timetableItemSummary.length, 'millimeters', 1),
            mechanicalEnergyConsumed: jouleToKwh(timetableItemSummary.energy_consumption, true),
            pathItemTimes: {
              base: timetableItemSummary.path_item_times_base,
              provisional: timetableItemSummary.path_item_times_provisional,
              final: timetableItemSummary.path_item_times_final,
            },
          }
        : {
            isValid: false,
            arrivalTime: null,
            duration: null,
            pathLength: '',
            mechanicalEnergyConsumed: 0,
            invalidReason:
              timetableItemSummary.status === 'pathfinding_not_found' ||
              timetableItemSummary.status === 'pathfinding_input_error'
                ? timetableItemSummary.error_type
                : timetableItemSummary.status,
          };

    const formattedItem = {
      ...timetableItem,
      name: timetableItem.train_name,
      startTime,
      stopsCount:
        (timetableItem.schedule?.filter(
          (step) => step.stop_for && Duration.parse(step.stop_for).ms > 0
        ).length ?? 0) + 1, // +1 to take the final stop (destination) into account
      speedLimitTag: timetableItem.speed_limit_tag ?? null,
      labels: timetableItem.labels ?? [],
      rollingStock,
      scheduledPointsNotHonored: notHonoredReason !== undefined,
      notHonoredReason,
      ...otherProps,
    };

    if (isPacedTrainResultWithPacedTrainId(formattedItem)) {
      return {
        ...formattedItem,
        paced: {
          duration: Duration.parse(formattedItem.paced.duration),
          step: Duration.parse(formattedItem.paced.step),
        },
      };
    }
    return formattedItem;
  });

  return mapBy(compact(items), 'id');
};
```

</details>

Does this `map()` approach not work on your machine?

SharglutDev replied:
Oh yes you're right. Not sure what I did before that made typescript complain.

emersion commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimatableItems.ts:
```diff
@@ -0,0 +1,162 @@
+          [[], []]
+        );
+
+        const rawTrainScheduleSummaries = await postTrainScheduleSimulationSummary({
```
Nit: would be slightly nicer to fetch these in parallel, without waiting for the train schedule request to complete before sending the paced train request.

```ts
const trainSchedulesPromise = postTrainScheduleSimulationSummary(…);
const pacedTrainsPromise = postPacedTrainSimulationSummary(…);

const rawTrainScheduleSummaries = await trainSchedulesPromise;
const rawPacedTrainSummaries = await pacedTrainsPromise;
```

SharglutDev replied:
Done

emersion commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimatableItems.ts:
```diff
@@ -0,0 +1,162 @@
+        const rawTrainScheduleSummaries = await postTrainScheduleSimulationSummary({
+          body: {
+            infra_id: _infraId,
+            ids: editoastTrainScheduleIds,
```
The list can be empty, would be nicer to skip sending an HTTP request when there is nothing to fetch.

SharglutDev replied:
Done

emersion reviewed (commented):
Here's another round of reviews :P

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
-      const trainIds = trainSchedules.map((trainSchedule) => trainSchedule.id);
+    if (timetableItems && projectionPath?.path && allTimetableItemsLoaded) {
+      // TODO Paced train : Adapt this to handle paced trains in issue https://github.com/OpenRailAssociation/osrd/issues/10613
+      const trainIds = timetableItems.map((timetableItem) => timetableItem.id);
```
Hum should you take all the trains here only for the moment ? 
```suggestion
      const trainIds = timetableItems
        .filter((timetableItem) => isTrainSchedule(timetableItem.id))
        .map((timetableItem) => timetableItem.id);
```

SharglutDev replied:
You're right. Done.

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
     }
   }, [projectionPath?.path]);
 
+  // TODO Paced trains : adapt this to handle paced trains in the conflicts issue
```
I think you can remove this comment ;) 
```suggestion
```

SharglutDev replied:
We didn't take care of the conflicts atm, there is still a pending [PR](https://github.com/OpenRailAssociation/osrd/pull/11069) for the paced train conflict endpoint

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
     setTrainIdsToProject,
   });
 
+  // TODO Paced trains : create a new useLazyProjectTimetableItems in https://github.com/OpenRailAssociation/osrd/issues/10613
```
Hm you will have to adapt the existing useLazyProject hook ?
```suggestion
```

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
-  }, [allTrainsLoaded]);
+  }, [allTimetableItemsLoaded]);
 
+  // TODO Paced trains : remove this hook in https://github.com/OpenRailAssociation/osrd/issues/10791
```
Maybe it's better like this ?
```suggestion
  // TODO Paced trains : adapt this hook in https://github.com/OpenRailAssociation/osrd/issues/10791
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
-      trainSchedules,
-      setTrainIdsToProject,
+      timetableItemIdsToFetch,
+      timetableItems,
```
Hum you have deleted the setTrainIdsToProject here, should it be replaced by the setTimetableItemsToProject ?

SharglutDev replied:
Oh you're right, as the original plan was to duplicated the hook, I forgot to put it back now that we merged everything. I also removed the above TODO.

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -12,30 +12,37 @@ import {
-  const [trainSchedules, setTrainSchedules] = useState<TrainScheduleResultWithTrainId[]>();
-  const [trainIdsToFetch, setTrainIdsToFetch] = useState<TrainId[]>();
-  const [trainIdsToProject, setTrainIdsToProject] = useState<Set<TrainId>>(new Set());
+  // TODO Paced trains : remove these 3 states in https://github.com/OpenRailAssociation/osrd/issues/10791
```
```suggestion
  // TODO Paced trains : remove this state in https://github.com/OpenRailAssociation/osrd/issues/10791
```

SharglutDev replied:
Removed the comment and renamed `trainIdsToProject` to `timetableItemIdsToProject`

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -12,30 +12,37 @@ import {
 
 import useAutoUpdateProjection from './useAutoUpdateProjection';
-import useLazyLoadTrains from './useLazyLoadTrains';
+import useLazyLoadTimetableItems from './useLazyLoadTimatableItems';
```
Typo :)
```suggestion
import useLazyLoadTimetableItems from './useLazyLoadTimetableItems';
```

SharglutDev replied:
Good catch !

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
-    () => sortBy(Array.from(trainScheduleSummariesById.values()), 'startTime'),
-    [trainScheduleSummariesById]
-  );
+  const timetableItemWithDetails = useMemo(() => {
```
```suggestion
  const timetableItemsWithDetails = useMemo(() => {
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
   );
 
-  const trainsIds = useMemo(() => trainSchedulesResults.map((t) => t.id), [trainSchedulesResults]);
+  const trainScheduleIds = useMemo(
```
```suggestion
  const trainScheduleEditoastIds = useMemo(
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
-        ...trainSchedule,
-        id: formatEditoastTrainIdToTrainScheduleId(trainSchedule.id),
-      })
+    const sortedTimatableItems = sortBy(
```
```suggestion
    const sortedTimetableItems = sortBy(
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
 
-      const sortedTrainSchedulesToUpsert = sortBy(trainSchedulesToUpsert, 'start_time');
-      setTrainIdsToFetch(sortedTrainSchedulesToUpsert.map((trainSchedule) => trainSchedule.id));
+      const sortedTimatableItemsToUpsert = sortBy(timetableItemsToUpsert, 'start_time');
```
```suggestion
      const sortedTimetableItemsToUpsert = sortBy(timetableItemsToUpsert, 'start_time');
```

SharglutDev replied:
Done 🙃

clarani commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimatableItems.ts:
```diff
@@ -0,0 +1,162 @@
+ * This hook gradually fetches and projects items of the timetable.
+ *
+ * It first fetches the simulation of 10 items at a time, then projects them on the path.
+ * This optimizes the performance of the application and allow us to display the items as
```
```suggestion
 * This optimizes the performance of the application and allows us to display the items as
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimatableItems.ts:
```diff
@@ -0,0 +1,162 @@
+  electricalProfileSetId?: number;
+  timetableItemIdsToFetch?: TimetableItemId[];
+  timetableItems?: TimetableItemWithTimetableId[];
+  setTimetableItemIdsToFetch?: Dispatch<SetStateAction<TimetableItemId[] | undefined>>;
```
I think this one is not used
```suggestion
```

SharglutDev replied:
Good catch

clarani commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimatableItems.ts:
```diff
@@ -0,0 +1,162 @@
+  const { data: { results: rollingStocks } = { results: null } } =
+    osrdEditoastApi.endpoints.getLightRollingStock.useQuery({ pageSize: 1000 });
+
+  const timetalbeItemsById = useMemo(() => mapBy(timetableItems, 'id'), [timetableItems]);
```
```suggestion
  const timetableItemsById = useMemo(() => mapBy(timetableItems, 'id'), [timetableItems]);
```

clarani commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimatableItems.ts:
```diff
@@ -0,0 +1,162 @@
+            }
+            return acc;
+          },
+          [[], []]
```
```suggestion
           { trainScheduleIds: [], pacedTrainIds: [] }
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimatableItems.ts:
```diff
@@ -0,0 +1,162 @@
+    ) => {
+      setAllTimetableItemsLoaded(false);
+
+      for (let i = 0; i < _timetableItemIdsToFetch.length; i += BATCH_SIZE) {
```
Hum do you think it would be better to fetch first all the paced Trains and then all the trainschedules (or the reverse ?) 
Otherwise, we might send twice more requests (if in each package we have both trainSchedules and pacedTrains, we might send each time a request for 5 trainSchedules and a request for 5 pacedTrains)

SharglutDev replied:
Good idea ! In the `useEffect` l.171, I changed the logic to put the train schedules and the the paced trains.

clarani commented on front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx:
```diff
@@ -103,10 +103,12 @@ const ScenarioContent = ({
       netzgrafikDto,
       addUpsertedTrainSchedules: (upsertedTrainSchedules: TrainScheduleResultWithTrainId[]) => {
-        upsertTrainSchedules(upsertedTrainSchedules);
+        // TODO Paced trains : update this in https://github.com/OpenRailAssociation/osrd/issues/10612
```
You can remove the 2 comments, since there is nothing to change here (everything will by in the functions ;) )
```suggestion
```

SharglutDev replied:
Done

Maymanaf commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
```diff
@@ -371,14 +407,148 @@ class ScenarioTimetableSection extends CommonPage {
+    await this.verifyOccurrencesCount(totalOccurrences, index);
+
+    await expect(this.hideOccurrencesButton.nth(index)).not.toBeVisible();
+    await expect(pacedTrainItem.locator('.toggle-icon')).toBeVisible();
```
why not use data-testid

SharglutDev replied:
Not sure why but with the `data-testid`, playwright doesn't find the `Locator` 🤔

Maymanaf replied:
[it's related to the React componenet rendering](https://legacy.reactjs.org/docs/forwarding-refs.html?) , you just can declare the data-testid inside a span, For exemple : 
```
      <span data-testid="hide-occurrences-button">
              <ChevronDown className="toggle-icon center-icon" />
            </span>
```
If you find this approach suboptimal, you can keep the current locator

Maymanaf commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
```diff
@@ -371,14 +407,148 @@ class ScenarioTimetableSection extends CommonPage {
+    // An invalid paced train won't have any details
+    if (labels?.includes('Invalid')) return;
+
+    const formattedDuration = dayjs.duration(pacedTrainDuration).asMinutes();
+    const formattedStep = dayjs.duration(step).asMinutes();
```
it's better to use the values directly

SharglutDev replied:
Done

Uriel-Sautron commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +60,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
+    (timetableItemsToUpsert: TimetableItemWithTimetableId[]) => {
+      // TODO Paced train : Add logic for projected timetable items in https://github.com/OpenRailAssociation/osrd/issues/10613
       setProjectedTrainsById((prev) => {
         const newProjectedTrainsById = new Map(prev);
-        trainSchedulesToUpsert.forEach((trainSchedule) => {
+        timetableItemsToUpsert.forEach((trainSchedule) => {
           newProjectedTrainsById.delete(trainSchedule.id);
         });
         return newProjectedTrainsById;
       });
```
```
      setProjectedItemsById((prev) => {
        const newProjectedItemsById = new Map(prev);
        timetableItemsToUpsert.forEach((timetableItem) => {
          newProjectedTrainsById.delete(timetableItem.id);
        });
        return newProjectedItemsById;
      });
```

SharglutDev replied:
I kept `trainSchedule` inside for now to remember that we manipulate only train schedule and not paced train. I planned that we would change this when dealing with projection in the related issue. Is that ok for you ?

Maymanaf commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
```diff
@@ -371,14 +407,148 @@ class ScenarioTimetableSection extends CommonPage {
+
+    const pacedTrainCadenceLocator = pacedTrainItem.getByTestId('paced-train-cadence');
+    await expect(pacedTrainCadenceLocator).toBeVisible();
+    await expect(pacedTrainCadenceLocator).toHaveText(
+      `${String.fromCodePoint(0x2014)} ${formattedStep}min`
+    ); // UI format: "- 20min"
```
you can directly check for 20min

SharglutDev replied:
Hmm the displayed text is "- 20min" so I have to deal with the special character no ? Or you want me to test a `.toEqual('20min')` and pass this hardcoded value when calling the function ?

Maymanaf replied:
`toHaveText() `check if 20min is in the text so the format is not needed
If you keep the format change it to` toEqual()`

SharglutDev replied:
Seems like he wants the exact match : 
<img width="774" alt="Capture d’écran 2025-03-14 à 16 00 09" src="https://github.com/user-attachments/assets/9a317aed-ee01-4999-9508-43529dbe6f57" />

Maymanaf replied:
My bad, `toContainText()` should work instead

Maymanaf commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
```diff
@@ -371,14 +407,148 @@ class ScenarioTimetableSection extends CommonPage {
   }
 
+  async verifyPacedTrainItemDetails(
+    pacedTrainData: Pick<PacedTrainBase, 'train_name' | 'start_time' | 'labels' | 'paced'>,
```
you can avoid using the project types and create a new type for the tests

SharglutDev replied:
Done

Maymanaf commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
```diff
@@ -371,14 +407,148 @@ class ScenarioTimetableSection extends CommonPage {
+    } = pacedTrainData;
+
+    const pacedTrainItem = this.timetableTrains.nth(pacedTrainIndex);
+    const occurrenceItem = pacedTrainItem.getByTestId('occurrence-item').nth(occurrenceIndex);
```
```suggestion
  getOccurrenceItem(occurrenceIndex: number): Locator {
    return this.page.getByTestId('occurrence-item').nth(occurrenceIndex);
  }
```

Maymanaf commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
```diff
@@ -371,14 +407,148 @@ class ScenarioTimetableSection extends CommonPage {
+
+    await expect(occurrenceItem.locator('.rolling-stock img')).toBeVisible();
+
+    await occurrenceItem.hover();
+    await expect(occurrenceItem.getByTestId('project-item')).not.toBeVisible();
+    await expect(occurrenceItem.getByTestId('duplicate-item')).not.toBeVisible();
+    await expect(occurrenceItem.getByTestId('edit-item')).not.toBeVisible();
+    await expect(occurrenceItem.getByTestId('delete-item')).not.toBeVisible();
```
```suggestion
  getOccurrenceItem(occurrenceIndex: number): Locator {
    return this.page.getByTestId('occurrence-item').nth(occurrenceIndex);
  }

  async getActionButtonsLocators(occurrenceIndex: number): Record<string, Locator> {
    const occurrenceItem = this.getOccurrenceItem(occurrenceIndex);
    await occurrenceItem.hover();
    return {
      projectItem: occurrenceItem.getByTestId('project-item'),
      duplicateItem: occurrenceItem.getByTestId('duplicate-item'),
      editItem: occurrenceItem.getByTestId('edit-item'),
      deleteItem: occurrenceItem.getByTestId('delete-item'),
    };
  }
  
   async verifyItemsNotVisible(occurrenceIndex: number): Promise<void> {
    const actionButtonsLocators = this.getActionButtonsLocators(occurrenceIndex);
    await Promise.all(
      Object.values(actionButtonsLocators).map(locator => expect(locator).not.toBeVisible())
    );
  }
```

SharglutDev replied:
> getOccurrenceItem(occurrenceIndex: number): Locator {
    return this.page.getByTestId('occurrence-item').nth(occurrenceIndex);
  }

This won't work because even if they aren't displayed, occurence items still exist in the DOM, so doing `page.get...` will always return the first group of occurrence. We need to locate them relatively to the current paced train we are testing.

Also regarding this [comment](https://github.com/OpenRailAssociation/osrd/pull/10966#discussion_r1991789818), It seems like you want me to fetch `occurrenceItem` only once but by calling it inside the `getActionButtonsLocators` method (and the one I'll be creating for name, startTime etc) means we will call it multiple times.

So instead of calling `getOccurrenceItem` in each method I could pass the current occurrence `Locator` we are looping on that we would fetch only once at the beginning of `verifyOccurrenceDetails`.

But if I do that (for example with the name) 
```
  async verifyOccurrenceName(currentPacedTrainLocator: Locator, expectedName: string) {
    const occurrenceNameLocator = currentPacedTrainLocator.locator('.occurrence-item-name');
    await expect(occurrenceNameLocator).toHaveText(expectedName);
  }
```

I have this eslint warning for each method : `Expected 'this' to be used by class async method 'verifyOccurrenceName'.`

So best solution seems to be the actual one. Or do you think of something else ?

Maymanaf commented on front/tests/assets/constants/operational-studies-const.ts:
```diff
@@ -1,22 +1,28 @@
+  labels: ['Valid'],
+  paced: {
+    duration: 'PT90M',
+    step: 'PT20M',
```
It better to hard code the values 90 and 20

Maymanaf commented on front/tests/pages/operational-studies/operational-studies-page.ts:
```diff
@@ -208,55 +216,75 @@ class OperationalStudiesPage extends CommonPage {
+    const formattedTimeRangeDuration = dayjs
+      .duration(DEFAULT_PACED_TRAIN_SETTINGS.paced.duration)
+      .asMinutes()
+      .toString();
```
Use hardcoded values

Maymanaf commented on front/tests/pages/operational-studies/operational-studies-page.ts:
```diff
@@ -208,55 +216,75 @@ class OperationalStudiesPage extends CommonPage {
+    const formattedCadence = dayjs
+      .duration(DEFAULT_PACED_TRAIN_SETTINGS.paced.step)
+      .asMinutes()
+      .toString();
```
same use hardcoded values

Maymanaf reviewed (commented):
Good job ! 
Left some comments

Maymanaf commented on front/tests/pages/operational-studies/scenario-timetable-section.ts:
Because multiple locators are scoped within occurrenceItem, you can move all of them in a new page model paced-train-section.ts
and it's best to store them in a reusable function so that you only fetch occurrenceItem once and reuse it

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -55,6 +59,7 @@ const SimulationResults = ({
   infraId,
   projectionData,
   trainScheduleSummaries,
+  // timetableItemsWithDetails,
```
```suggestion
```

SharglutDev replied:
Done

clarani commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -44,6 +45,33 @@ const osrdEditoastApi = generatedEditoastApi
+                { forceRefetch: true, subscribe: false }
+              )
+            );
+            // need to unsubscribe on get call to avoid cache issue
```
Hm have you unsubscribe here ? Or it is not necessary ?

SharglutDev replied:
Hmm I reproduced the way it is done for the above custom endpoint. I asked myself this question and thought the `subscribe: false` was doing the job. @emersion what about it ?

emersion replied:
Passing `{ subscribe: false }` as options will indeed avoid subscribing, thus there is no need to unsubscribe later.

IMHO the comment can be dropped.

clarani replied:
The comment is still here

emersion replied:
Comment should now be gone!

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/AddTrainScheduleButton.tsx:
```diff
@@ -51,21 +53,80 @@ const AddTrainScheduleButton = ({
-    if (validTrainConfig) {
-      const { timetableId, firstStartTime, trainCount, trainDelta, trainStep, baseTrainName } =
-        validTrainConfig;
+    const { timetableId, firstStartTime, trainCount, trainDelta, trainStep, baseTrainName } =
```
You should setIsWorking(true) here

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/AddTrainScheduleButton.tsx:
```diff
@@ -51,21 +53,80 @@ const AddTrainScheduleButton = ({
+    const { timetableId, firstStartTime, trainCount, trainDelta, trainStep, baseTrainName } =
+      validTimetableItemConfig;
 
+    if (showPacedTrains) {
+      try {
```
Move the try above the test showPacedTrains
```suggestion
    try {
      if (showPacedTrains) {
      ...
```

SharglutDev replied:
I kept it like this to facilitate the moment we will have to delete the `trainSchedule` code in the `else` instead of merging all try/catch. Is that ok fo you ?

clarani replied:
Ok :)

SharglutDev committed:
front: fix weird behavior when adding train settings

This extra useEffect introduced a bug in end to end tests. The only way to reproduce it was to set the train name and start time in less that 500ms (the debounce used for these fields) which the e2e tests do.
At this speed, after filling the name, a useEffect is fired after 500ms after to update the store, but since the test is also filling the start time a few ms later, the useEffect was called because nameFromStore had been updated resulting in setting the start time back at its original value.
A workaround had been done for tests 011 and 012 to insert the selection of the rolling stock between the name and start time filling.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: fix paced train action buttons display

The buttons were displayed when hovering the paced train item and the occurrences item when it should be only on paced train hovering.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion commented on front/tests/utils/date-utils.ts:
```diff
@@ -44,6 +44,9 @@ export function getLocalizedDateString(dateString: string): string {
 export const createDateInSpecialTimeZone = (dateString: string, timeZone: string) =>
   dayjs.tz(dateString, timeZone);
 
+export const formatDayJSDateToHHMM = (dateString: string, timeZone: string) =>
+  dayjs(dateString).tz(timeZone).format('HH:mm');
```
Why not hardcode the result in test data instead?

SharglutDev replied:
This won't scale when the time zone changes since we have some times 1 or 2 hours diff with UTC

emersion replied:
Since everything is relative to `start_time`, it should be fine!

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -4,44 +4,149 @@ import { Checkbox } from '@osrd-project/ui-core';
+      body: { ids: [formatPacedTrainIdToEditoastTrainId(pacedTrain.id)] },
+    })
+      .unwrap()
+      .then(() => {
```
We can use `await` here instead of dealing with callbacks.

SharglutDev replied:
Done

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -4,44 +4,149 @@ import { Checkbox } from '@osrd-project/ui-core';
+      id: editoastTrainId,
+    })
+      .unwrap()
+      .catch((e) => {
```
We can use `try`/`catch` here instead of dealing with callbacks.

SharglutDev replied:
Done

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -4,44 +4,149 @@ import { Checkbox } from '@osrd-project/ui-core';
+        dispatch(setFailure(castErrorToFailure(e)));
+      });
+
+    if (pacedTrainDetail) {
```
An early return can be used instead of deeply nesting the rest of the function. But that should be unnecessary if we `return` in a `catch` block above.

SharglutDev replied:
Done

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -4,44 +4,149 @@ import { Checkbox } from '@osrd-project/ui-core';
+
+    const editoastTrainId = formatPacedTrainIdToEditoastTrainId(pacedTrain.id);
+    const pacedTrainDetail = await getPacedTrainById({
+      id: editoastTrainId,
```
This subscribes to the query without ever unsubscribing. This means the paced train will be kept in the cache forever until the user refreshes the page.

This can be avoided by passing `{ subscribe: false }` as options.

SharglutDev replied:
I don't think `lazyQuery` accepts `subscribe` as options like when using `initiate` ([docs](https://redux-toolkit.js.org/rtk-query/api/created-api/hooks#uselazyquery-signature)).
I used `.unsubscribe()` to do it manually.

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -4,44 +4,149 @@ import { Checkbox } from '@osrd-project/ui-core';
+        const [pacedTrainResult] = await postPacedTrain({
+          id: pacedTrainDetail.timetable_id,
+          body: [newPacedTrain],
+        }).unwrap();
```
Nit: no need to nest everything here in the `try` block.

```ts
let pacedTrainResult;
try {
  [pacedTrainResult] = await postPacedTrain({
    id: pacedTrainDetail.timetable_id,
    body: [newPacedTrain],
  }).unwrap();
} catch (e) {
  dispatch(setFailure(castErrorToFailure(e)));
  return;
}

// rest of the code
```

SharglutDev replied:
Done

emersion commented on front/src/applications/operationalStudies/hooks/useLazyLoadTimetableItems.ts:
```diff
@@ -0,0 +1,181 @@
+                  electrical_profile_set_id: electricalProfileSetId,
+                },
+              }).unwrap()
+            : ({} as Promise<PostTrainScheduleSimulationSummaryApiResponse>);
```
This cast sounds dangerous.

`Promise.resolve({})` should work I think?

SharglutDev replied:
Done

emersion commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -53,158 +61,194 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
-
-  // first load of the trainScheduleSummaries
+    const sortedTimetableItems = [
+      ...sortBy(formattedRawTrainSchedules, 'start_time'),
+      ...sortBy(formattedRawPacedTrains, 'start_time'),
```
It seems like this puts first all train schedules in the list, and then all paced trains?

Shouldn't we sort the whole list by `start_time`?

SharglutDev replied:
See this [comment](https://github.com/OpenRailAssociation/osrd/pull/10966#discussion_r1991060923) :)

emersion replied:
Ah yeah, this is only for fetching, not for displaying. Makes sense!

clarani commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -205,41 +247,79 @@ const TimetableToolbar = ({
     return `${pacedTrainLabel}\u00A0${t('translation:common.and')}\u00A0${trainScheduleLabel}`;
   };
 
+  // TODO : doesn't work for german. Find a better way to handle this
+  const deleteTimetableItemsComputedLabel = () => {
```
This function can be moved in the `DeleteModal`

SharglutDev replied:
Done

emersion commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -205,41 +247,79 @@ const TimetableToolbar = ({
+    }
+
+    if (selectedPacedTrainIds.length > 0 && selectedTrainScheduleIds.length > 0) {
+      itemsLabel = `${pacedTrainLabel}\u00A0${t('translation:common.and')}\u00A0${trainScheduleLabel}`;
```
This assumes a particular grammatical structure of the language. This structure will not work for some languages.

Can we instead use nesting? See:
https://www.i18next.com/translation-function/nesting

SharglutDev replied:
Yeah... I put a comment for that at the top of the function. Now that we have the german language, should we also update it each time we add/change a translation file ?

SharglutDev replied:
Ok done, was easier than I thought

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -145,14 +164,37 @@ const TimetableToolbar = ({
 
-    // TODO Paced train : Adapt this to handle export paced trains in issue https://github.com/OpenRailAssociation/osrd/issues/10614
-    const formattedTrainSchedules = trainSchedules
+    const formattedTrainSchedules = timetableItems
```
Should you add a small filter here to keep only the trainSchedules and not the paced trains ?

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -166,8 +208,8 @@ const TimetableToolbar = ({
-  const computedItemLabel = (trainSchedulesCount: number, pacedTrainCount: number) => {
-    if (trainSchedulesCount === 0 && pacedTrainCount === 0) return t('timetable.noItem');
+  const computedItemLabel = () => {
+    if (totalPacedTrainCount === 0 && totalPacedTrainCount === 0) return t('timetable.noItem');
```
```suggestion
    if (totalPacedTrainCount === 0 && totalTrainScheduleCount === 0) return t('timetable.noItem');
```

SharglutDev replied:
Oups done

clarani commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -180,14 +222,14 @@ const TimetableToolbar = ({
 
     if (
-      trainSchedulesCount === 0 ||
+      totalPacedTrainCount === 0 ||
```
```suggestion
      totalTrainScheduleCount === 0 ||
```

SharglutDev replied:
Oups again

clarani commented on front/src/modules/trainschedule/components/Timetable/TimetableToolbar.tsx:
```diff
@@ -205,41 +247,79 @@ const TimetableToolbar = ({
+
+    let itemsLabel = '';
+
+    if (
+      totalPacedTrainCount === 0 ||
+      (selectedPacedTrainIds.length > 0 && selectedTrainScheduleIds.length === 0)
+    ) {
+      itemsLabel = pacedTrainLabel;
+    }
+
+    if (
+      totalPacedTrainCount === 0 ||
+      (selectedTrainScheduleIds.length > 0 && selectedPacedTrainIds.length === 0)
+    ) {
+      itemsLabel = trainScheduleLabel;
+    }
+
+    if (selectedPacedTrainIds.length > 0 && selectedTrainScheduleIds.length > 0) {
+      itemsLabel = `${pacedTrainLabel}\u00A0${t('translation:common.and')}\u00A0${trainScheduleLabel}`;
+    }
```
You don't need to test the total counts here, only the selected counts matter. You can simplify the code like this:
```suggestion
    if (selectedPacedTrainIds.length > 0 && selectedTrainScheduleIds.length > 0) {
      itemsLabel = `${pacedTrainLabel}\u00A0${t('translation:common.and')}\u00A0${trainScheduleLabel}`;
    } else {
      itemsLabel = selectedPacedTrainIds.length > 0 ? pacedTrainLabel : trainScheduleLabel;
    }

```

SharglutDev replied:
After a comment from @emersion I changed this function a bit. Let me know if it's ok for you.

clarani replied:
Yes perfect ! I hesitated to ask the same thing :star_struck:

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/PacedTrainSettings.tsx:
```diff
@@ -9,13 +11,22 @@ import {
-  const timeRangeDuration = useSelector(getTimeRangeDuration);
-  const cadence = useSelector(getCadence);
+  const timeRangeDurationStore = useSelector(getTimeRangeDuration);
+  const cadenceStore = useSelector(getCadence);
```
Hum do you think we can do this ?
```suggestion
  const timeRangeDurationStore = useSelector(getTimeRangeDuration).total('minute');
  const cadenceStore = useSelector(getCadence).total('minute');
```

SharglutDev replied:
Looks like it's working yes !

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -4,55 +4,168 @@ import { Checkbox } from '@osrd-project/ui-core';
   selectedTimeTableItemId: TrainId | undefined;
+  selectPacedTrainToEdit: (pacedTrain: PacedTrainWithDetails) => void;
+  upsertTimetableItems: (timetableItems: TimetableItemWithTimetableId[]) => void;
+  removePaceTrains: (pacedTrainIdsToRemove: TimetableItemId[]) => void;
```
```suggestion
  removePacedTrains: (pacedTrainIdsToRemove: TimetableItemId[]) => void;
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -0,0 +1,104 @@
+
+    const timetableItemSummary = rawSummaries.get(timetableItem.id);
+
+    if (!timetableItemSummary) return null;
```
This should never happen since you create relevantTimetableItems from rawTimetableItems. You can throw an Exception here
```suggestion
    if (!timetableItemSummary) {
      throw new Error(...)
    }
```

SharglutDev replied:
Done

clarani commented on front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -0,0 +1,104 @@
+                : timetableItemSummary.status,
+          };
+
+    const formattedItem = {
+      ...timetableItem,
+      name: timetableItem.train_name,
+      startTime,
+      stopsCount:
+        (timetableItem.schedule?.filter(
+          (step) => step.stop_for && Duration.parse(step.stop_for).ms > 0
+        ).length ?? 0) + 1, // +1 to take the final stop (destination) into account
+      speedLimitTag: timetableItem.speed_limit_tag ?? null,
+      labels: timetableItem.labels ?? [],
+      rollingStock,
+      scheduledPointsNotHonored: notHonoredReason !== undefined,
+      notHonoredReason,
+      ...otherProps,
+    };
+
+    if (isPacedTrainResultWithPacedTrainId(formattedItem)) {
+      return {
+        ...formattedItem,
+        paced: {
+          duration: Duration.parse(formattedItem.paced.duration),
+          step: Duration.parse(formattedItem.paced.step),
+        },
+      };
+    }
+    return formattedItem;
```
```suggestion
    return {
      ...timetableItem,
      name: timetableItem.train_name,
      startTime,
      stopsCount:
        (timetableItem.schedule?.filter(
          (step) => step.stop_for && Duration.parse(step.stop_for).ms > 0
        ).length ?? 0) + 1, // +1 to take the final stop (destination) into account
      speedLimitTag: timetableItem.speed_limit_tag ?? null,
      labels: timetableItem.labels ?? [],
      rollingStock,
      scheduledPointsNotHonored: notHonoredReason !== undefined,
      notHonoredReason,
      ...otherProps,
      ...(isPacedTrainResultWithPacedTrainId(timetableItem)
        ? {
            paced: {
              duration: Duration.parse(timetableItem.paced.duration),
              step: Duration.parse(timetableItem.paced.step),
            },
          }
        : {}),
    };
```

SharglutDev replied:
Hmm I have a typescript error when I try to do this (because `timetableItem` type is a bit weird)

clarani replied:
OK, let's keep it like this :)

emersion reviewed (approved):
Overall LGTM! Very nice work :)

clarani reviewed (approved):
LGTM :white_check_mark:

Maymanaf commented on front/tests/005-operational-studies.spec.ts:
```diff
@@ -103,19 +127,45 @@ test.describe('Verify simulation configuration in operational studies for train
   // TODO Paced train : Remove this test in https://github.com/OpenRailAssociation/osrd/issues/10791
+  /** *************** Test 3 **************** */
   test('Pathfinding with rolling stock and composition code', async ({ page }) => {
     // Page models
 
```
The first part is already called in the beforeEach

SharglutDev replied:
Done

Maymanaf commented on front/tests/pages/common-page.ts:
```diff
@@ -31,9 +31,18 @@ class CommonPage {
 
-  // Verify the text of the last toast notification title
+  // Verify the text of all toast notification titles
   async checkLastToastTitle(expectedText: string | RegExp): Promise<void> {
```
```suggestion
  async checkToastTitle(expectedText: string | RegExp): Promise<void> {
```

SharglutDev replied:
Done

Maymanaf commented on front/tests/008-timetable-item.spec.ts:
```diff
@@ -242,4 +251,80 @@ test.describe('Verify train schedule elements and filters', () => {
+      './tests/assets/paced-train/paced_trains.json'
+    );
+
+    // Verify paced train item
+    for (let i = 0; i < pacedTrainsData.length; i += 1) {
+      await pacedTrainSection.verifyPacedTrainItemDetails(
+        IMPORTED_PACED_TRAIN_DETAILS[i],
+        i,
+        IMPORT_PACED_TRAIN_OCCURRENCES_DETAILS[i]
+      );
+    }
```
```suggestion
    // Verify paced train item
    for (let pacedTrainIndex = 0; i < pacedTrainsData.length; pacedTrainIndex += 1) {
      await pacedTrainSection.verifyPacedTrainItemDetails(
        IMPORTED_PACED_TRAIN_DETAILS[pacedTrainIndex],
        pacedTrainIndex
        IMPORT_PACED_TRAIN_OCCURRENCES_DETAILS[pacedTrainIndex]
      );
    }
```

SharglutDev replied:
Done

Maymanaf commented on front/tests/008-timetable-item.spec.ts:
```diff
@@ -242,4 +251,80 @@ test.describe('Verify train schedule elements and filters', () => {
+    }
+  });
+
+  /** *************** Test 5 **************** */
```
move this test to 005-operational-studies.spec.ts

SharglutDev replied:
Done

Maymanaf commented on front/tests/016-timetable-multiselection.spec.ts:
This test is flaky 
![image](https://github.com/user-attachments/assets/24842011-c64e-4151-84cb-d1d398bac577)

SharglutDev replied:
Since our recent changes, can't reproduce the flaky test anymore (test 016)

Maymanaf commented on front/tests/pages/operational-studies/paced-train-section.ts:
```diff
@@ -0,0 +1,242 @@
+
+    await this.verifyItemsVisibility(index, 'paced-train');
+
+    for (let i = 0; i < totalOccurrences; i += 1) {
+      await this.verifyOccurrenceDetails(occurrenceData[i], i, {
+        copyTranslation: duplicate?.copyTranslation,
+      });
+    }
+
```
```suggestion
    for (let occurrenceIndex = 0; i < totalOccurrences; occurrenceIndex += 1) {
      await this.verifyOccurrenceDetails(occurrenceData[occurrenceIndex], occurrenceIndex, {
        copyTranslation: duplicate?.copyTranslation,
      });
    }

```

SharglutDev replied:
Done

Maymanaf commented on front/tests/pages/operational-studies/paced-train-section.ts:
```diff
@@ -0,0 +1,242 @@
+    translations: TimetableFilterTranslations
+  ) {
+    const duplicatedPacedTrainName = `${deletedPacedTrainName} (${translations.timetable.copy})`;
+    // translation has format 'La mission {{name}} a bien été supprimée';
```
```suggestion
    // Translation has format 'The service {{name}} has been deleted';
```

SharglutDev replied:
Done

Maymanaf reviewed (approved):
LGTM for e2e tests ✅
